java/flight/flight-core/src/main/java/org/apache/arrow/flight/ClientApiCallWrapper.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.flight;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import org.apache.arrow.flight.grpc.StatusUtils;
+
+import io.grpc.stub.StreamObserver;
+
+/**
+ * Utility class for executing an API calls on the FlightServer.
+ */
+public class ClientApiCallWrapper {
+
+  /**
+   * Execute client call using the Stream Observer ApiCallObserver.
```

Comment:
nit: more specifically, this converts an async callback-based call into a synchronous iterator